### PR TITLE
luci-proto-gre: improvements of interface

### PR DESCRIPTION
**Pull request by mbargo23**
Code has been modified to better handling of the protocol settings (especially handling of ToS and TC values).
When creating a new interface, the optional value specifying the local endpoint address is pre-filled in the form, based on the current address of wan interface.

### CODE REVIEW

**Comment by feckert** [comment]
@mbargo23  you please rebase your changes

**Comment by mbargo23** [comment]
> @mbargo23 you please rebase your changes

I'm really sorry for this, but how can I do that?

**Comment by mbargo23** [comment]
> @mbargo23 you please rebase your changes

Fixed, thank you.

**Comment by jow-** [comment]
Merged, thanks!

**Comment by oofnikj** [comment]
Thanks for this! I got it working in 19.07 just by copying the protocol definitions to `/www/luci-static/resources/protocol/`.
Any chance we can have this officially backported to the latest stable release?

**Comment by feckert** [comment]
If you want this in openwrt stable, then you need to do a PR with the following commits.
https://github.com/openwrt/luci/commit/2ced86048c7006f3637ed38ceeb3d21fddbb93e2#diff-cf5fa1724cdf261be6b497865374713926de614facb9efac70f864db6454d0cd
https://github.com/openwrt/luci/commit/2b7fd1292f3c5055e9f5c431e411c53d3b84adf5#diff-cf5fa1724cdf261be6b497865374713926de614facb9efac70f864db6454d0cd
https://github.com/openwrt/luci/commit/37f97b7859c8a601024ae6570e9d481c50ee8897#diff-cf5fa1724cdf261be6b497865374713926de614facb9efac70f864db6454d0cd
https://github.com/openwrt/luci/commit/0f3f464adc9e24646638aa5d71edb995a352dec5#diff-cf5fa1724cdf261be6b497865374713926de614facb9efac70f864db6454d0cd

**Comment by mbargo23** [comment]
Thanks! I will handle it.

**Comment by mbargo23** [comment]
> Thanks! I will handle it.

I'm giving up learning how to work with git properly.
Can someone create the PR instead of me, please?

**Comment by oofnikj** [comment]
@mbargo23 A little late to the party but I finally got around to it.

> I'm giving up learning how to work with git properly.

Don't give up! Git is many things, but easy to use is not one of them. The process basically consisted of cherry-picking the commits identified by @feckert and handling the merge conflicts that arose due to some translations missing in 19.07. This was simplified by passing the following arguments: `git cherry-pick --strategy=recursive -X theirs ...` where `...` is the commit hash you want to merge in to the current branch.

In plain English, this means Git will recursively try to merge incoming changes, favoring the incoming change over existing commits in the case of a conflict instead of leaving the user to decide the outcome. Next step is to open a PR with the cherry-picked commits and hope it gets accepted :)

**Comment by mbargo23** [comment]
> Don't give up! Git is many things, but easy to use is not one of them. The process basically consisted of cherry-picking the commits identified by @feckert and handling the merge conflicts that arose due to some translations missing in 19.07. This was simplified by passing the following arguments: `git cherry-pick --strategy=recursive -X theirs ...` where `...` is the commit hash you want to merge in to the current branch.
> 
> In plain English, this means Git will recursively try to merge incoming changes, favoring the incoming change over existing commits in the case of a conflict instead of leaving the user to decide the outcome. Next step is to open a PR with the cherry-picked commits and hope it gets accepted :)

@oofnikj I will try it again :-) Thank you for support.
